### PR TITLE
refactor: add error codes to API error callbacks

### PR DESCRIPTION
**Pull request by DanDeMicco**
This is a follow up to #650. Mainly it adds error codes and additional flow typing to the API layer for 90% of API's which were sometimes added in components in the previous PR

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @DanDeMicco has signed the CLA. Thanks for the pull request!

**Comment by boxcla** [comment]
# Webpack Bundle Size Github Bot

| Bundle        | Diff | Percentage Diff | Before | After |
| ------------- |:----:|:---------------:|:------:|:-----:|
| Total | 0 B | 0.00% | 0 B | 0 B |

Compared against previously merged PR #651

**Powered by [webpack-bundle-size-github-bot](https://github.com/wenboyu2/webpack-bundle-size-github-bot/)**

File: src/api/Base.js [hunk 1]
```diff
@@ -349,8 +354,8 @@ class Base {
                 ...requestData,
             });
             this.successHandler(data);
-        } catch (error) {
-            this.errorHandler(error);
+        } catch (e) {
```
Comment (jstoffan) [inline src/api/Base.js]: Curious for the reason behind this change? I think we usually want to avoid abbreviations.

File: src/api/Comments.js [hunk 2]
```diff
@@ -5,7 +5,17 @@
  */
 
 import OffsetBasedAPI from './OffsetBasedAPI';
-import { PERMISSION_CAN_COMMENT, PERMISSION_CAN_DELETE, PERMISSION_CAN_EDIT } from '../constants';
+import {
+    PERMISSION_CAN_COMMENT,
+    PERMISSION_CAN_DELETE,
+    PERMISSION_CAN_EDIT,
+    ERROR_CODE_CREATE_COMMENT,
```
Comment (jstoffan) [inline src/api/Comments.js]: Would be great to put these behind a namespace, like `ERROR_CODE.CREATE_COMMENT` or `PERMISSION.CAN_COMMENT`

File: src/api/Comments.js [hunk 3]
```diff
@@ -229,17 +242,21 @@ class Comments extends OffsetBasedAPI {
...
         try {
             this.checkApiCallValidity(PERMISSION_CAN_COMMENT, permissions, fileId);
         } catch (e) {
-            errorCallback(e);
+            errorCallback(e, this.errorCode);
             return;
         }
 
-        this.offsetGet(fileId, successCallback, errorCallback, ...rest);
+        this.offsetGet(fileId, successCallback, errorCallback, offset, limit, fields, shouldFetchAll);
```
Comment (jstoffan) [inline src/api/Comments.js]: Is this change related to the error codes? It seems like the defaults should live in the parent class rather than being duplicated in each child.

File: src/api/Feed.js [hunk 4]
```diff
@@ -182,94 +170,66 @@ class Feed extends Base {
...
-                (e: ElementsXhrError) => {
-                    this.fetchFeedItemErrorCallback(e, errorCallback, ERROR_CODE_FETCH_VERSIONS);
-                    resolve();
-                },
-                DEFAULT_START,
-                DEFAULT_END,
-                VERSIONS_FIELDS_TO_FETCH,
-                true,
-            );
+            this.versionsAPI.getVersions(this.id, resolve, this.fetchFeedItemErrorCallback.bind(this, resolve));
```
Comment (jstoffan) [inline src/api/Feed.js]: It seems odd that the inner methods would resolve successfully after an error. Ideally, they would reject and the parent should catch and handle them appropriately.

File: src/api/Feed.js [hunk 5]
```diff
@@ -297,6 +257,7 @@ class Feed extends Base {
         }
 
         this.id = file.id;
+        this.onError = errorCallback;
```
Comment (jstoffan) [inline src/api/Feed.js]: In some places we assign this value from the `onError` param and in others from the `errorCallback` param. It's not immediately clear what the distinction is or when to use one or the other.

File: src/api/Folder.js [hunk 6]
```diff
@@ -315,7 +320,9 @@ class Folder extends Item {
                 },
             })
             .then(this.createSuccessHandler)
-            .catch(this.errorHandler);
+            .catch((e: $AxiosError<any>) => {
```
Comment (jstoffan) [inline src/api/Folder.js]: What's the value of wrapping this in an anonymous function?

File: src/api/Base.js (hunk 1 above)
Comment (DanDeMicco) [reply src/api/Base.js]: No particular reason. I saw e being used in a few places, so just changed it to be consistent

File: src/api/Comments.js (hunk 3 above)
Comment (DanDeMicco) [reply src/api/Comments.js]: No not really, I can remove some of the defaults and just make them optional

File: src/api/Feed.js (hunk 4 above)
Comment (DanDeMicco) [reply src/api/Feed.js]: Agreed. The reason why they resolve undefined is that `Promise.all` is being used to fetch the comments, tasks, versions. If we reject, we cannot show partial data as it will hit the .catch.

A custom promise.all could be implemented (or imported) to handle this. Another unfortunate thing about reject is that it can only return 1 value, but I need 2 (error and code from the error callback). We have an eslint rule that enforces it to be of type Error. So, I could create an Error object with the code stuck into it.

File: src/api/Feed.js (hunk 5 above)
Comment (DanDeMicco) [reply src/api/Feed.js]: Would adding JSDoc to the `onError` property on the class help? Maybe changing the variable name?

```
     * @param {Function} errorCallback - the error callback which is called after data fetching is complete if there was an error
     * @param {Function} onError - the function to be called immediately after an error occurs
```

File: src/api/Folder.js (hunk 6 above)
Comment (DanDeMicco) [reply src/api/Folder.js]: I think this is leftover from a refactor of this where I was before passing the error callback to the errorHandler, but then changed it to be a class property
